File: practice_handout/sorting/merge_sort.js
var mergeSort = function(array) {
  if (array.length <= 1) {return array};
  var middle = array.length / 2;
  leftHalf = mergeSort(array.slice(0, middle));
  rightHalf = mergeSort(array.slice(middle, array.length));
  console.log('left, right')
  console.log(leftHalf)
  console.log(rightHalf)
  return merge(leftHalf, rightHalf);
}

var merge = function(a,b) {
  var merged = [];
  while (a.length > 0 && b.length > 0) {
    if (a[0] < b[0]) {
      merged.push(a.shift());
    }
    else {
      merged.push(b.shift());
    }
  }
  while (a.length > 0) {
    merged.push(a.shift());
  }
  while (b.length > 0) {
    merged.push(b.shift());
  }
  return merged;
}

console.log(mergeSort([1,2,3,4,5]))//===[1,2,3,4,5])
// console.log(mergeSort([1,5,4,2,3]))//===[1,2,3,4,5])

File: practice_handout/sorting/merge_sort.rb
# takes an array and returns a sorted array
def merge_sort(array)
  return array if array.length <= 1
  middle = array.length / 2
  first_half = merge_sort(array[0..middle-1])
  second_half = merge_sort(array[middle..-1])
  merge(first_half,second_half)
end

def merge(a,b)
  merged_array = []
  until a.empty? || b.empty?
    a.first < b.first ? merged_array << a.shift : merged_array << b.shift
  end
  (b.empty? ? a : b).each { |x| merged_array << x }
  merged_array
  # merged_array.push(b.empty? ? a : b).flatten
end

def other_sort(array)
  array.sort {|a,b| a <=> b}
end



# Driver code
puts (merge_sort([1,2,3,4]) == [1,2,3,4])
puts (merge_sort([3,4,2,1]) == [1,2,3,4])
puts (merge_sort([3,4,2,2,2,1]) == [1,2,2,2,3,4])
shuffled_array = (1..100).to_a.shuffle
puts merge_sort(shuffled_array) == (1..100).to_a
long_shuffled = (1..100000).to_a.shuffle
puts merge_sort(long_shuffled) == (1..100000).to_a








# start = Time.now
# puts (merge_sort(shuffled_array) == (1..100).to_a)
# puts "100 numbers takes #{Time.now - start} with merge sort"
# long_shuffled = (1..100000).to_a.shuffle
# start = Time.now
# puts (merge_sort(long_shuffled) == (1..100000).to_a)
# puts "100000 numbers takes #{Time.now - start} with merge sort"

# start = Time.now
# puts (other_sort(shuffled_array) == (1..100).to_a)
# puts "100 numbers takes #{Time.now - start} with bubble sort"
# long_shuffled = (1..100000).to_a.shuffle
# start = Time.now
# puts (other_sort(long_shuffled) == (1..100000).to_a)
# puts "100000 numbers takes #{Time.now - start} with bubble sort"

